File: backend/dotnet/Services/TaskService.cs
using dotnet.Data;
using dotnet.exceptions;
using dotnet.Interfaces;
using dotnet.Models;
using Task = dotnet.Models.Task;

namespace dotnet.Services;

public class TaskService : ITaskService
{
    private readonly DataContext _context;
    private readonly IUserService _userService;

    public TaskService(DataContext context, IUserService userService)
    {
        _context = context;
        _userService = userService;
    }

    public IEnumerable<Task> GetAllTasks()
    {
        return _context.Tasks.ToList();
    }

    public IEnumerable<Task> GetAllTasksForProject(string projectId)
    {
        return _context.Tasks.Where(t => t.ProjectId == projectId).ToList();
    }

    public void CreateTask(Task task)
    {
        _context.Tasks.Add(task);
        _context.SaveChanges();
    }

    public void UpdateTask(string name, string description, bool finished, string taskId)
    {
        var task = _context.Tasks.FirstOrDefault(t => t.Id == taskId) ??
                   throw new CustomException("Task doesn't exist", 400);

        task.Name = name;
        task.Description = description;
        task.Finished = finished;
        task.FinishedBy = finished ? _userService.GetCurrentUser().Name : null;

        _context.Update(task);
        _context.SaveChanges();
    }

    public void DeleteTask(string id)
    {
        var task = _context.Tasks.FirstOrDefault(t => t.Id == id) ??
                   throw new CustomException("Task doesn't exist", 400);

        _context.Tasks.Remove(task);
        _context.SaveChanges();
    }

    public void AddCommentToTask(string taskId, string comment)
    {
        TaskComment taskComment = new TaskComment()
        {
            Task = _context.Tasks.Where(t => t.Id == taskId).FirstOrDefault(),
            Content = comment,
            CreatedAt = DateTime.Now,
            User = _userService.GetCurrentUser(),
            TaskId = taskId,
            UserId = _userService.GetCurrentUser().Email
        };
        _context.TaskComments.Add(taskComment);
        _context.SaveChanges();
    }

    public void StarringTask(string taskId, string userEmail)
    {
        var task = _context.Tasks.FirstOrDefault(t => t.Id == taskId) ??
                   throw new CustomException("Task doesn't exist", 400);

        var user = _userService.GetUserByEmail(userEmail);

        var existingStar = _context.StarredTasks
            .FirstOrDefault(st => st.Task == task && st.User == user);

        if (existingStar != null)
        {
            _context.StarredTasks.Remove(existingStar);
        }
        else
        {
            var starredTask = new StarredTask()
            {
                Task = task,
                User = user,
                CreatedAt = DateTime.Now
            };
            _context.StarredTasks.Add(starredTask);
        }

        _context.SaveChanges();
    }
}

File: backend/springboot/src/main/java/tn/maiko26/springboot/service/TaskService.java
package tn.maiko26.springboot.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tn.maiko26.springboot.exception.CustomException;
import tn.maiko26.springboot.model.Task;
import tn.maiko26.springboot.model.relations.StarredTask;
import tn.maiko26.springboot.model.relations.TaskComment;
import tn.maiko26.springboot.repository.StarredTaskRepository;
import tn.maiko26.springboot.repository.TaskCommentRepository;
import tn.maiko26.springboot.repository.TaskRepository;

import java.util.Date;
import java.util.List;
import java.util.Optional;

@Service
public class TaskService {

    @Autowired
    private TaskRepository taskRepository;

    @Autowired
    private TaskCommentRepository taskCommentRepository;

    @Autowired
    private StarredTaskRepository starredTaskRepository;
    @Autowired
    private UserService userService;

    // Get all tasks
    public List<Task> getAllTasks() {
        return taskRepository.findAll(); // Assuming you have a repository for Task entity
    }

    // Get tasks by project
    public List<Task> getTasksForProject(String projectId) {
        return taskRepository.findByProjectId(projectId); // Query to get tasks by projectId
    }

    // Create a new task
    public Task createTask(Task task) {
        return taskRepository.save(task); // Save the task entity
    }

    // Update an existing task
    public Task updateTask(String name, String description, Boolean finished, String taskId) {
        Optional<Task> existingTask = taskRepository.findById(taskId);
        if (existingTask.isPresent()) {
            Task updatedTask = existingTask.get();
            updatedTask.setName(name);
            updatedTask.setDescription(description);
            updatedTask.setFinished(finished);
            updatedTask.setFinishedBy(finished ? userService.getUserByEmail(userService.getCurrentUserEmail()) : null);
            return taskRepository.save(updatedTask); // Save the updated task entity
        } else {
            throw new CustomException("Task doesn't exist", 400);
        }
    }

    // Delete task by id
    public void deleteTask(String taskId) {
        taskRepository.deleteById(taskId); // Delete the task by ID
    }

    // Add comment to a task
    public TaskComment addCommentToTask(TaskComment taskComment) {
        return taskCommentRepository.save(taskComment); // Save the task comment entity
    }

    // Star or unstar a task
    public void starringTask(String taskId, String userEmail) {
        Optional<StarredTask> existingStar = starredTaskRepository.findByTaskAndUser(taskRepository.findById(taskId).orElseThrow(() -> new CustomException("Couldn't find the task", 400)), userService.getUserByEmail(userEmail));

        if (existingStar.isPresent()) {
            starredTaskRepository.delete(existingStar.get()); // If already starred, unstar the task
        } else {
            StarredTask starredTask = new StarredTask();
            starredTask.setTask(taskRepository.findById(taskId).orElseThrow(() -> new CustomException("Task not found.", 400)));
            starredTask.setUser(userService.getUserByEmail(userEmail));
            starredTask.setCreatedAt(new Date());
            starredTaskRepository.save(starredTask); // Otherwise, star the task
        }
    }
}
